lib/middleware/request_tracker.rb

```diff
@@ -416,11 +420,32 @@ def find_auth_cookie(env)
     cookie if cookie && cookie[:issued_at] >= min_allowed_timestamp
   end
 
-  def is_private_ip?(ip)
-    ip = IPAddr.new(ip)
-    !!(ip && (ip.private? || ip.loopback?))
-  rescue IPAddr::AddressFamilyError, IPAddr::InvalidAddressError
-    false
+  def self.rate_limit_bucketizer_register(priority:, identifier:, blk:)
+    new_list = @@bucketizers.dup.delete_if { _1[1] == identifier }
+    new_list.unshift([priority, identifier, blk])
+    @@bucketizers = new_list.sort_by { _1[0] } # priority
+  end
+
+  def self.rate_limit_bucketizer_unregister(identifier:)
+    @@bucketizers.delete_if { _1[1] == identifier }
+  end
+
+  def rate_limit_bucket_for_request(request, cookie)
+    # returns: [
+    #   bucket,                   # the specific bucket value, i.e. IP address / username / something else
+    #   global?,                  # is this limiter global across sites?
+    #   short bucket description, # short type of the bucket, i.e. "ip", "id", "cloud"
+    #   long bucket description,  # type of the bucket, intended for human eyes
+    # ] | :skip
+
+    return :skip if @@ip_skipper&.call(request.ip)
```

Comment:
I think we should keep the API to skip rate limiting of a request as a separate API instead of supporting it as a part of the new API because it is technically two different concerns. One API is to meant to register conditions to bypass rate limiting entirely and another API is to register different conditions on how the rate limiting is to be carried out.

Reply:
We could do that.

When I first looked at the code and saw multiple methods for rate limit bypassing I figured we could simplify things by deprecating those entirely over time.

E.g. one of those was added a number of years ago solely for the ability to have a customer's plugin allowlist a set of IPs.

We could eliminate this entirely by adjusting that plugin to instead register a bypass function.

Arguably it's the same concern since a valid response to "which bucket" is "no bucket"